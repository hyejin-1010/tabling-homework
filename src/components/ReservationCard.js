import Component from '@/core/Component';
import timeFormat from '@/utils/timeFormat';

import './ReservationCard.scss';

const statusText = {
  reserved: '예약',
  seated: '착석 중'
}

export default class ReservationCard extends Component {
  // TODO: Card / Action Button Click Event

  template() {
    const { reservation } = this.props;

    return `
      <div class="card">
        <!-- 좌측 / 중앙 / 우측 : 3열로 구분하여 표출 -->

        <!--
          좌측 : 예약 시간과 상태를 1열 2행으로 중앙정렬로 표현
                예약 상태는 한글로 표출
                  - reserved : "예약" (#3BB94C)
                  - seated : "착성 중" (#162149)
        -->
        <div class="card-status">
          <!-- 예약 시간 -->
          <div>${timeFormat(reservation.timeReserved)}</div>
          <!-- 예약 상태 -->
          <div class="${reservation.status}">${statusText[reservation.status]}</div>
        </div>

        <!--
          중앙 : 예약자명, 테이블명, 방문자 수, 및 메뉴를 1열 3행으로 표출
                - 예약자명 및 테이블명은 아래 형식에 맞춰 한 줄로 표현
                  "예약자명 - 테이블명 [, 테이블명]"
                - 방문자수는 "성인 00 아이 00" 형식에 맞춰 표출
                - 메뉴는 "메뉴명(갯수)[, 메뉴명(갯수)]"
        -->
        <div class="card-content">
          <!-- 예약자명 및 테이블명 -->
          <div class="ellipsis">
            <span>${reservation.customer.name}</span>
            -
            <span>
              ${reservation.tables.map((table) => table.name).join(', ')}
            </span>
          </div>

          <!-- 방문자수 -->
          <div class="card-visitors">
            성인 ${reservation.customer.adult}
            아이 ${reservation.customer.child}
          </div>
          <!-- 메뉴 -->
          <div class="ellipsis">
            ${
              reservation.menus.map((menu) => {
                return `${menu.name}(${menu.qty})`;
              }).join(', ')
            }
          </div>
        </div>

        <!--
          우측 : [착석] 또는 [퇴석] 버튼을 표출
                - reserved 상태면 [착석] : 클릭 시, [퇴석] 으로 버튼을 변경
                - seated 상태면 [퇴석] : 클릭 시, 예약 목록에서 제거
        -->
        <div class="card-action">
          <button id="action-btn" class="action-btn ${reservation.status}">
            ${ reservation.status === 'reserved' ? '착석' : '퇴석' }
          </button>
        </div>
      </div>
    `;
  }
}
