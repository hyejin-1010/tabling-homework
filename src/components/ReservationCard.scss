.card {
  display: grid;
  grid-template-columns: 1fr 5fr 1fr;
  margin: 0.75rem 0;
  padding: 0.75rem 0;
  background: white;
  border-radius: 10px;
}

.card-status, .card-action {
  padding: 1.25rem;
}

.card-status {
  display: flex;
  flex-direction: column;
  justify-content: center;
  text-align: center;
  .reserved {
    color: #3BB94C;
  }
  .seated {
    color: #162149;
  }
}

.card-content {
  overflow: hidden;
  display: flex;
  flex-direction: column;
  justify-content: center;
  font-size: 0.75rem;
  .card-visitors {
    margin: 0.5rem 0;
  }
}

.card-action {
  align-self: center;
  .action-btn {
    width: 100%;
    padding: 0.5rem 0;
    background: #162149;
    border: none;
    border-radius: 5px;
    color: white;
    &.seated {
      background: white;
      border: 1px solid #162149;
      color: #162149;
    }
  }
}

// @media only screen and (max-width: 1024px) {
@media only screen and (max-width: 700px) {
  .card {
    grid-template-columns: 1fr 2fr 1fr;
  }
  .card-status, .card-action {
    padding: 0.75rem;
  }
}