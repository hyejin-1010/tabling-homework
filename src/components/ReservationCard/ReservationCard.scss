@import '@/styles/colors.scss';
@import '@/styles/variables.scss';

.card {
  display: grid;
  grid-template-columns: 1fr 4fr 1fr;
  margin: 0.85rem 0;
  padding: 0.85rem 1rem;
  background: $white;
  border-radius: 10px;
}

.card-left, .card-action {
  padding: 0.85rem;
}

.card-left {
  display: flex;
  flex-direction: column;
  justify-content: center;
  text-align: center;
  .status {
    margin-top: 0.5rem;
  }
  .reserved {
    color: $green;
  }
  .seated {
    color: $primary;
  }
}

.card-content {
  overflow: hidden;
  display: flex;
  flex-direction: column;
  justify-content: center;
  font-size: $font-size-sm;
  .card-visitors {
    margin: 0.5rem 0;
  }
}

.card-action {
  align-self: center;
  .action-btn {
    width: 100%;
    padding: 0.5rem 0;
    background: $primary;
    border: none;
    border-radius: 5px;
    color: $white;
    font-size: $font-size-base;
    &.seated {
      background: $white;
      border: 1px solid $primary;
      color: $primary;
    }
  }
}

@media only screen and (max-width: 1024px) {
  .card {
    grid-template-columns: 1fr 2fr 1fr;
    margin: 0.85rem 0;
    padding: 0.85rem 0.25rem;
  }
  .card-left, .card-action {
    padding: 0.5rem;
  }
  .card-action .action-btn {
    font-size: $font-size-sm;
  }
}