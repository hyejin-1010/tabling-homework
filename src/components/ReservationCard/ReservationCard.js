import Component from '@/core/Component';
import timeFormat from '@/utils/timeFormat';
import translate from '@/utils/translate';

import './ReservationCard.scss';

/**
 * @property {object} reservation 예약 정보
 * @property {function(reservation: object)} [onClick] 컴포넌트 클릭 시
 * @property {function(id: string)} [onSeat] 착석 버튼 클릭 시
 * @property {function(id: string)} [onDone] 퇴석 버튼 클릭 시
 */
export default class ReservationCard extends Component {
  didMount() {
    // 클릭 event 연결
    const { reservation, onClick, onSeat, onDone } = this.props;
    this.addEvent('click', '.card-content, .card-status', function() {
      if (onClick) { onClick(reservation); }
    });
    this.addEvent('click', '#action-btn.reserved', function() {
      if (onSeat) { onSeat(reservation.id); }
    });
    this.addEvent('click', '#action-btn.seated', function() {
      if (onDone) { onDone(reservation.id); }
    });
  }


  template() {
    const { reservation } = this.props;

    return `
      <div class="card">
        <div class="card-left">
          <!-- 예약 시간 -->
          <div>${timeFormat(reservation.timeReserved)}</div>
          <!-- 예약 상태 -->
          <div class="${reservation.status} status">${translate[reservation.status]}</div>
        </div>

        <div class="card-content">
          <!-- 예약자명 및 테이블명 -->
          <div class="ellipsis">
            <span>${reservation.customer.name}</span>
            <span>-</span>
            <span>${reservation.tables.map((table) => table.name).join(', ')}</span>
          </div>

          <!-- 방문자수 -->
          <div class="card-visitors">
            <span>성인 ${reservation.customer.adult}</span>
            <span>아이 ${reservation.customer.child}</span>
          </div>

          <!-- 메뉴 -->
          <div class="ellipsis">
            ${
              reservation.menus.map((menu) => {
                return `${menu.name}(${menu.qty})`;
              }).join(', ')
            }
          </div>
        </div>

        <!-- [착석] 또는 [퇴석] 버튼 -->
        <div class="card-action">
          <button id="action-btn" class="action-btn ${reservation.status}">
            ${reservation.status === 'reserved' ? '착석' : '퇴석'}
          </button>
        </div>
      </div>
    `;
  }
}
