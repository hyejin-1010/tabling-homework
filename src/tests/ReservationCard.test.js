import { describe, expect, test } from '@jest/globals';
import ReservationCard from '@/components/ReservationCard/ReservationCard';
import Data from './mock';

describe('ReservationCard Component Snapshot Test', () => {
  test('Basic Data', () => {
    const element = document.createElement('div');
    expect(new ReservationCard(element, { reservation: Data.reservation })).toMatchSnapshot();
  })
});
