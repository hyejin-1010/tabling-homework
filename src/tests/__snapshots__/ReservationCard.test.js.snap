// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReservationCard Component Snapshot Test Basic Data 1`] = `
ReservationCard {
  "props": {
    "reservation": {
      "customer": {
        "adult": 2,
        "child": 1,
        "id": "200512254359",
        "level": "VIP",
        "memo": "독일 태생의 이론물리학자로서 역사상 가장 위대한 물리학자 중의 한명으로 널리 인정되고 있다. 상대성 이론을 개발한 것으로 가장 잘 알려져 있지만 양자역학 이론의 발전에도 중요한 공헌을 했다.",
        "name": "알베르트 아인슈타인",
        "request": "-",
        "timeVisitedLast": "2020-09-11 18:32:20",
      },
      "id": "53005895",
      "menus": [
        {
          "id": "02110701",
          "name": "봉골라 스파게티",
          "qty": 1,
        },
        {
          "id": "02090105",
          "name": "파이브 치즈 피자",
          "qty": 1,
        },
      ],
      "status": "seated",
      "tables": [
        {
          "floor": 1,
          "id": 0,
          "max": 4,
          "min": 1,
          "name": "1번 테이블",
        },
      ],
      "timeRegistered": "2020-10-10 18:02:20",
      "timeReserved": "2020-11-02 11:00:00",
    },
  },
  "state": {},
  "target": <div>
    
      
    <div
      class="card"
    >
      
        
      <div
        class="card-left"
      >
        
          
        <!-- 예약 시간 -->
        
          
        <div>
          11:00
        </div>
        
          
        <!-- 예약 상태 -->
        
          
        <div
          class="seated status"
        >
          착석 중
        </div>
        
        
      </div>
      

        
      <div
        class="card-content"
      >
        
          
        <!-- 예약자명 및 테이블명 -->
        
          
        <div
          class="ellipsis"
        >
          
            
          <span>
            알베르트 아인슈타인
          </span>
          
            
          <span>
            -
          </span>
          
            
          <span>
            1번 테이블
          </span>
          
          
        </div>
        

          
        <!-- 방문자수 -->
        
          
        <div
          class="card-visitors"
        >
          
            
          <span>
            성인 2
          </span>
          
            
          <span>
            아이 1
          </span>
          
          
        </div>
        

          
        <!-- 메뉴 -->
        
          
        <div
          class="ellipsis"
        >
          
            봉골라 스파게티(1), 파이브 치즈 피자(1)
          
        </div>
        
        
      </div>
      

        
      <!-- [착석] 또는 [퇴석] 버튼 -->
      
        
      <div
        class="card-action"
      >
        
          
        <button
          class="action-btn seated"
          id="action-btn"
        >
          
            퇴석
          
        </button>
        
        
      </div>
      
      
    </div>
    
    
  </div>,
}
`;
